### PR TITLE
adding top-down error oversight

### PR DESCRIPTION
**Pull request by 11chankun**
adding top-down error oversight along with the data file required for this and the test file. I shall add some other test file after some time. I need some suggestion regarding the oversight suggestion string.

### CODE REVIEW

File: intents/oversights/top_down_error.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+    Returns:
+        return the similarity between both the arguments calculated by the
+        formula
+        similarity = |val1 - val2| / (|val1| + |val2|)
+    """
+
+    if val1 == val2:
+        return 0
+    result = abs(val1 - val2) / (abs(val1) + abs(val2))
```
Comment (bhagyakjain) [inline intents/oversights/top_down_error.py:245]: Duplication of code. A better check would be -
if (abs(val1) + abs(val2)) == 0 :
    return 0

File: intents/oversights/top_down_error.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        new_suggestion['is_row_level_suggestion'] = True
+        new_suggestion['row_list'] = suggestion_row_list
+        return new_suggestion
+
+def _calculate_relation(val1, val2):
+    """
+    This function can find the similarity between two values
+    """
+
+    """
```
Comment (bhagyakjain) [inline intents/oversights/top_down_error.py:232]: Remove these 3 lines

File: intents/oversights/top_down_error.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    if len(suggestion_row_list) == 0:
+        return None
+    else:
+        new_suggestion = {}
+        new_suggestion['suggestion'] = 'both the value is almost same in some grouping but different if we add ' + new_added_column + ' in initial grouping.'
+        new_suggestion['oversight_name'] = 'top-down error'
+        new_suggestion['is_row_level_suggestion'] = True
+        new_suggestion['row_list'] = suggestion_row_list
+        return new_suggestion
```
Comment (bhagyakjain) [inline intents/oversights/top_down_error.py:225]: add key 'confidence_score' if feasible

File: intents/oversights/top_down_error.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+    # dominant percentage is the percentage of pairs where 
+    # (first value - second value) is positive.
+
+    while initial_row_i < initial_num_rows:
+        if initial_row_i == initial_num_rows - 1 or initial_table_matrix[initial_row_i][:dimensions_len] != initial_table_matrix[initial_row_i + 1][:dimensions_len]:
+            while new_row_i < new_num_rows and initial_table_matrix[initial_row_i][:dimensions_len] == new_table_matrix[new_row_i][:dimensions_len]:
+                new_row_i = new_row_i + 1
+        else:
+            new_max_corelation = 0
```
Comment (bhagyakjain) [inline intents/oversights/top_down_error.py:198]: corelation -> correlation

File: intents/oversights/top_down_error.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            # initial result table and new result table.
+            new_suggestion =  _check_top_down_error(initial_result_table,
+                                                      new_result_table, 
+                                                      column, slice1,
+                                                      dimensions_len)
+            if new_suggestion != None:
+                suggestions.append(new_suggestion)
+    return suggestions
+
+def _check_top_down_error(initial_result_table, new_result_table, new_added_column,
```
Comment (bhagyakjain) [inline intents/oversights/top_down_error.py:159]: "initial_result_table" would the result table change ?

File: intents/oversights/top_down_error.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        return the similarity between both the arguments calculated by the
+        formula
+        similarity = |val1 - val2| / (|val1| + |val2|)
+    """
+
+    if val1 == val2:
+        return 0
+    result = abs(val1 - val2) / (abs(val1) + abs(val2))
+
+    return result
```
Comment (bhagyakjain) [inline intents/oversights/top_down_error.py:247]: Add end of line

File: intents/test_all.sh [hunk 7]
```diff
@@ -91,6 +91,14 @@ else
...
 fi
 
+if ! python3 oversights/test_top_down_error.py;
+then
+  err "oversights/test_top_down_error failed"
+  list_logs+=("oversights/test_top_down_error failed")
+else
+  list_logs+=("oversights/test_top_down_error passed")
+fi
+
```
Comment (bhagyakjain) [inline intents/test_all.sh:101]: Some of the tests are failing - 
1. Debiasing sentence was only updated in the code & not in assertions
2. Merge conflicts in constants.py

File: intents/util/constants.py [hunk 8]
```diff
@@ -39,3 +39,9 @@
 
 # threshold of disimilarity of two values in benchmark set too different.
 BSTD_DISIMILARITY_THRESHOLD = 0.20
+
+# top-down error similarity threshold value
+TOP_DOWN_ERROR_SIMILARITY_THRESHOLD = 0.05
+
+# top-down error dissimilarity threshold value
+TOP_DOWN_ERROR_DISSIMILARITY_THRESHOLD = 0.30
```
Comment (bhagyakjain) [inline intents/util/constants.py:47]: Add end of line

File: intents/data/student_score2.csv [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+7th,A,english,30
+7th,B,english,90
+8th,B,maths,90
+8th,A,maths,90
+8th,B,Social science,90
+8th,A,Social science,90
+8th,B,science,90
+8th,A,science,90
+8th,B,english,90
+8th,A,english,90
```
Comment (bhagyakjain) [inline intents/data/student_score2.csv:17]: add end of line

File: intents/oversights/top_down_error.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+            if new_max_corelation >= constants.TOP_DOWN_ERROR_DISSIMILARITY_THRESHOLD and initial_corelation <= constants.TOP_DOWN_ERROR_SIMILARITY_THRESHOLD:
+                suggestion_row_list.append(initial_row_i)
+                suggestion_row_list.append(initial_row_i + 1)
+            initial_row_i = initial_row_i + 1
+
+        initial_row_i = initial_row_i + 1
+
+    if len(suggestion_row_list) == 0:
+        return None
```
Comment (bhagyakjain) [inline intents/oversights/top_down_error.py:218]: return None -> return

File: intents/oversights/top_down_error.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+    # dominant percentage is the percentage of pairs where 
+    # (first value - second value) is positive.
+
+    while initial_row_i < initial_num_rows:
+        if initial_row_i == initial_num_rows - 1 or initial_table_matrix[initial_row_i][:dimensions_len] != initial_table_matrix[initial_row_i + 1][:dimensions_len]:
+            while new_row_i < new_num_rows and initial_table_matrix[initial_row_i][:dimensions_len] == new_table_matrix[new_row_i][:dimensions_len]:
+                new_row_i = new_row_i + 1
+        else:
+            new_max_corelation = 0
+            while new_row_i < new_num_rows and initial_table_matrix[initial_row_i][:dimensions_len] == new_table_matrix[new_row_i][:dimensions_len]:
```
Comment (bhagyakjain) [inline intents/oversights/top_down_error.py:199]: Use \ to break lines.

File: intents/oversights/top_down_error.py (hunk 3 above)
Comment (11chankun) [reply intents/oversights/top_down_error.py:225]: I haven't think about it. so I won't be able to add it right now.

File: intents/oversights/top_down_error.py (hunk 11 above)
Comment (bhagyakjain) [reply intents/oversights/top_down_error.py:199]: Example - https://github.com/googleinterns/debaised-analysis/blob/master/date_detection/date_column_type_detection.py#L41

File: intents/oversights/top_down_error.py (hunk 1 above)
Comment (11chankun) [reply intents/oversights/top_down_error.py:245]: Done

File: intents/oversights/top_down_error.py (hunk 2 above)
Comment (11chankun) [reply intents/oversights/top_down_error.py:232]: Done

File: intents/oversights/top_down_error.py (hunk 4 above)
Comment (11chankun) [reply intents/oversights/top_down_error.py:198]: Done, Thanks

File: intents/oversights/top_down_error.py (hunk 6 above)
Comment (11chankun) [reply intents/oversights/top_down_error.py:247]: Done

File: intents/test_all.sh (hunk 7 above)
Comment (11chankun) [reply intents/test_all.sh:101]: Resolved this already

File: intents/util/constants.py (hunk 8 above)
Comment (11chankun) [reply intents/util/constants.py:47]: Done

File: intents/data/student_score2.csv (hunk 9 above)
Comment (11chankun) [reply intents/data/student_score2.csv:17]: I am not sure how to do this for data file

File: intents/oversights/top_down_error.py (hunk 10 above)
Comment (11chankun) [reply intents/oversights/top_down_error.py:218]: done

File: intents/oversights/top_down_error.py (hunk 11 above)
Comment (11chankun) [reply intents/oversights/top_down_error.py:199]: will do thanks

File: intents/data/student_score2.csv (hunk 9 above)
Comment (bhagyakjain) [reply intents/data/student_score2.csv:17]: Simply go to that line & press enter.

**Comment by bhagyakjain** [comment]
Overall logic looks good.

File: intents/oversights/top_down_error.py (hunk 11 above)
Comment (bhagyakjain) [reply intents/oversights/top_down_error.py:199]: @11chankun  your forked is 1 commit behind the googleinterns:master , just update it with googleinterns. 
Thanks

Comment (11chankun) [reply intents/oversights/top_down_error.py:199]: thanks done pull

File: intents/data/student_score2.csv (hunk 9 above)
Comment (11chankun) [reply intents/data/student_score2.csv]: done

File: intents/oversights/top_down_error.py (hunk 3 above)
Comment (11chankun) [reply intents/oversights/top_down_error.py]: I have added confidence score as 100 right now.
It might be changed.

**Pull request merged**